sources/packages/backend/apps/workers/src/services/disbursement-schedule/disbursement-schedule-service.ts

```diff
@@ -1,28 +1,86 @@
 import { Injectable } from "@nestjs/common";
-import { DataSource, Repository } from "typeorm";
+import {
+  DataSource,
+  EntityManager,
+  In,
+  IsNull,
+  MoreThanOrEqual,
+  Repository,
+} from "typeorm";
 import {
   RecordDataModelService,
   DisbursementSchedule,
   DisbursementValue,
   StudentAssessment,
   AssessmentTriggerType,
   ApplicationStatus,
+  DisbursementValueType,
+  DisbursementScheduleStatus,
+  DisbursementOveraward,
+  DisbursementOverawardOriginType,
+  Student,
+  Application,
+  configureIdleTransactionSessionTimeout,
 } from "@sims/sims-db";
-import { Disbursement } from "./disbursement-schedule.models";
+import {
+  DisbursementSaveModel,
+  DisbursementScheduleValue,
+} from "./disbursement-schedule.models";
 import { CustomNamedError } from "@sims/utilities";
 import {
   ASSESSMENT_INVALID_OPERATION_IN_THE_CURRENT_STATE,
   ASSESSMENT_NOT_FOUND,
   DISBURSEMENT_SCHEDULES_ALREADY_CREATED,
 } from "../../constants";
+import { QueryDeepPartialEntity } from "typeorm/query-builder/QueryPartialEntity";
+import { DisbursementOverawardService } from "..";
+import { SystemUsersService } from "@sims/services/system-users";
+
+// Timeout to handle the worst-case scenario where the commit/rollback
+// was not executed due to a possible catastrophic failure.
+const TRANSACTION_IDLE_TIMEOUT_SECONDS = 600;
+
+const LOAN_TYPES = [
+  DisbursementValueType.CanadaLoan,
+  DisbursementValueType.BCLoan,
+];
+
+const GRANTS_TYPES = [
+  DisbursementValueType.CanadaGrant,
+  DisbursementValueType.BCGrant,
+];
 
 /**
  * Service layer for Student Application disbursement schedules.
+ * Assumptions and concepts:
+ * - sims.disbursement_overawards can also be referred as overaward balance table or overaward history table.
+ * - One assessment/reassessment will have one or two dates to be disbursed, a.k.a. disbursement schedule.
+ * - Awards are the loans/grants associated with one disbursement schedule (e.g. CSLF, CSPT, BCSL).
+ * - The same award can be present in one or two schedules.
+ * - A student debt can be a value already received by the student for the same application or prior application.
+ *     - Values already paid(sent/disbursed): sum of all the awards already disbursed for that application number grouped by its code value (e.g. CSLF, CSPT, BCSL).
+ *     - Overaward balance: value from a prior application that must be deducted from the awards, if possible.
+ *     - During a reassessment, the new award calculated will be deducted from all the values already paid for that particular
+ * application. If any award is still to be received, the process will check for possible overaward balance to be deducted.
+ * - If there is a student debt, the priority is to deduct the value as much as possible from the sooner disbursement schedule.
+ * This means that, in case there are two disbursements for the same award, the process will try to deduct as much as possible
+ * from the first disbursement.
```

Comment:
and if there is still dept, the remaining from the upcoming/ 2nd disbursement right?

Reply:
Yes, try to get as much as possible from the first one, then from the second one, then, if there is a left over and it is a loan, an overaward is generated.